### ISSUE TITLE
DDC (and others?): Type cast failures should include <dynamic>

### ISSUE THREAD

**Issue by matanlurey** [dart-lang/sdk#32302]
For example, a popular case error at runtime looks like this:

```
Cast failure from JSArray to List<ExplicitType>
```

One reason this is extra confusing to me (and users), is that lack of an explicit type (and implicit dynamic type arguments) are not shown. For example, the following in DDC:

```dart
void main() {
  print([].runtimeType);
}
```

... prints `List`, not `List<dynamic>`.

Applying a fix here would make the above cast failures nicer too:

```
Cast failure from JSArray<dynamic> to List<ExplicitType>
```

/cc @nshahan @vsmenon

**Comment by vsmenon** [comment]
Some variants:
- Always show `Object instead of `dynamic`.  They are essentially the same thing (with fuzzy arrows gone.
- Always show `List` instead of `JSArray`.

Thoughts?

@jmesserly

**Comment by jmesserly** [comment]
Testing the different implementations, first for explicit casts:
```dart
main() {
  [] as int;
}
```
```
# VM output:
type 'List' is not a subtype of type 'int' in type cast where
  List is from dart:core
  int is from dart:core

# dart2js output:
CastError: Instance of 'JSArray': type 'JSArray' is not a subtype of type 'int'

# dartdevc output:
CastError: Casting value of type 'JSArray' to type 'int' which is incompatible
```
Feels like VM message is most helpful. DDC's message wording presumably is to accommodate the suffix "which is incompatible in strong mode", but I don't like it.

Next up is implicit casts (`int x = [] as dynamic;`). This now shows:
```
# VM output:
type 'List' is not a subtype of type 'int' of 'x' where
  List is from dart:core
  int is from dart:core

# dart2js output:
TypeError: Instance of 'JSArray': type 'JSArray' is not a subtype of type 'int'

# dartdevc output:
Type 'JSArray' is not a subtype of type 'int'
```
DDC's wording is a bit better, but it would be nice to show `List` instead of `JSArray`.

I don't think `Type.toString()` is specified, but conventionally it does not print dynamic type arguments. Perhaps it should, especially since the dynamic/Object distinction is not especially meaningful at runtime in Dart 2. I'd worry about changing that in general. We could special case it in DDC's TypeError/CastError toString.

**Comment by matanlurey** [comment]
Wow, thanks for the deep dive!!

/cc @leafpetersen @lrhn what do you think about what Dart 2 should do here?

**Comment by jmesserly** [comment]
Question for @leafpetersen @lrhn ... is `dynamic` vs `Object` reified distinctly in Dart 2?

Here are some examples:
```dart
class C<T> { Type get t => T; }
Type f<T>() => T;
T g<T>(T t) => t;

main() {
  print(dynamic); // dynamic
  print(Object);  // Object
  print(identical(dynamic,  Object)); // false

  var cd = new C();
  var co = new C<Object>();
  print(cd.t); // dynamic 
  print(co.t); // Object
  print(cd.runtimeType == co.runtimeType); // false

  var fd = f<dynamic>();
  var fo = f<Object>();
  print(fd); // dynamic
  print(fo); // Object

  Function(dynamic) gd = g;
  Object Function(Object) go = g;
  print(gd.runtimeType); // (dynamic) => dynamic
  print(go.runtimeType); // VM is (dynamic) => dynamic, DDC is (Object) => Object
  print(gd.runtimeType == go.runtimeType); // VM true, DDC false
}
class C<T> { Type get t => T; }
Type f<T>() => T;
T g<T>(T t) => t;

main() {
  print(dynamic); // dynamic
  print(Object);  // Object
  print(identical(dynamic,  Object)); // false

  var cd = new C();
  var co = new C<Object>();
  print(cd.t); // dynamic 
  print(co.t); // Object
  print(cd.runtimeType == co.runtimeType); // false

  var fd = f<dynamic>();
  var fo = f<Object>();
  print(fd); // dynamic
  print(fo); // Object

  Function(dynamic) gd = g;
  Object Function(Object) go = g;
  print(gd.runtimeType); // (dynamic) => dynamic
  print(go.runtimeType); // (Object) => Object
  print(gd.runtimeType == go.runtimeType); // false
}
```

**Comment by jmesserly** [comment]
(I fixed duplicate code pasted above. Apologies!)

**Comment by leafpetersen** [comment]
For the purposes of runtime `is` and `as` checks, `void`, `dynamic` and `Object` are indistinguishable.

It's probably better for users that they get reified separately since error messages will correspond more closely to source code.

I don't think the spec currently defines equality/identity on runtime type objects, but maybe it does?  

I think the least surprising is to reify them separately, and make them non equal as runtime type objects.

**Comment by eernstg** [comment]
About the distinctions between `void` and other top types, we do have the following in the feature specification [generalized-void.md](https://github.com/dart-lang/sdk/blob/master/docs/language/informal/generalized-void.md):

> For a composite type (a generic class instantiation or a function type),
> the reified representation at run time must be such that the type void and
> the built-in class `Object` are treated as equal according to `==`, but
> they need not be `identical`.
> ...
>
> *In case of a dynamic error, implementations are encouraged to emit an
> error message that includes information about such parts of types being
> `void` rather than `Object`. Developers will then see types which are
> similar to the source code declarations. This may be achieved using
> distinct `Type` objects to represent types such as `F<void, void>` and
> `F<Object, void>`, comparing equal using `==` but not `identical`.*

I do not think the language team has discussed the relationship between `dynamic` and other top types from a similar point of view, but obviously Dart 1 required `dynamic` to be considered distinct from "other top types" (it's not so obvious that `dynamic` is a top type in Dart 1), so there's an historic reason why `dynamic == Object` yields false.

The consistent approach might be to say that `dynamic == Object` should yield true, and only `identical` (and `toString` and diagnostic messages) should be allowed to distinguish the two. But it's not obvious to me that the improved consistency is worth the potential breakage in this case, and it would certainly not be a developer-friendly move to erase the difference completely (such that even error messages etc. can't see the difference).

**Comment by lrhn** [comment]
The specification doesn't specify equality of `Type` objects, but just that they "represent types".
We have generally enforced that `Type` objects are equal if they represent the same type.

In Dart 2, that might not be enough because we have multiple equivalent types, and `Object`, `void`, `dynamic` are not the problem (I expect those to *be* the same type at runtime, even if their `toString` differs, so equal-but-not-identical).

We have a type like `FutureOr<Object>` which is both a super-type and sub-type of `Object`, but is not the same type. We should probably extend equality on `Type` objects to equate those two, so `==` just means `<:` and `:>`.

We already knew that `Type` objects were not canonicalized, so they already need to override `==`. The ones that are compile-time constant are canonicalized, but anything created at runtime need not be. There can be arbitrarily many different types available at runtime, and requiring canonicalization would be a potential memory leak (especially in JS compiled code where we can't use GC-aware caching).

**Comment by jmesserly** [comment]
https://dart-review.googlesource.com/c/sdk/+/52081

**Comment by matanlurey** [comment]
@jmesserly any update here?

FWIW (and I +'d you on it) internal users are hitting a variant of these quite commonly:

```dart
abstract class BaseFoo<A, B> {}
class ConcreteFoo<T> extends BaseFoo<String, T> {}
class AnotherType {}

void main() {
  // Will be inferred as ConcreteFoo<dynamic>
  var foo = new ConcreteFoo();

  // Will fail. We passed a BaseFoo<String, dynamic> instead of <String, AnotherType>.
  useFoo(foo);
}

void useFoo(BaseFoo<String, AnotherType> foo) {}
```

In this instance, the error message they receive is:

```
Type 'ConcreteFoo' is not a subtype of `BaseFoo<String, AnotherType>` in Strong Mode.
```

It would really preferred here to, at minimum, highlight the implicit `<dynamic>`:

```
Type 'ConcreteFoo<dynamic>' is not a ...
```

An even better error would show the discrepancy directly:

```
Type 'ConcreteFoo<dynamic>' implements 'BaseFoo<String, dynamic>', but 'BaseFoo<String, AnotherType>' was expected.
```

/cc @vsmenon

**Comment by jmesserly** [comment]
@matanlurey -- the `<dynamic>` and `JSArray` will be fixed once my CL above lands.

I'll open another issue for the feature request of showing the discrepancy. It will take some new runtime machinery that we don't currently have available (not hard hard implement, just not something that currently exists in the runtime type system). But I like the idea. :)

**Comment by jmesserly** [comment]
the other issue: https://github.com/dart-lang/sdk/issues/32980
